File: src/trl/opt.rs
#[derive(Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Debug, Hash)]
pub enum Opt<T> {
    None,
    Some(T),
}

pub use Opt::None;
pub use Opt::Some;

impl<T> Opt<T> {}

File: src/trl/opt.ts
export enum OptType {
    Ok,
    No,
}

export interface Ok<T> {
    type: OptType.Ok
    val: T
}

export interface No<T = void> {
    type: OptType.No
    val: T
}

export class Opt<Y, N = void> {
    val: Ok<Y> | No<N>

    constructor(val: Ok<Y> | No<N>) {
        this.val = val
    }
}

export const Ok = <Y, N = void>(val: Y): Opt<Y, N> => {
    return new Opt<Y, N>({
        type: OptType.Ok,
        val
    })
}

export const No = <Y, N>(val?: N): Opt<Y, N> => {
    return new Opt<Y, N>({
        type: OptType.No,
        val: val!
    })
}
